### PR TITLE
Implement Oracle contract

### PR DESCRIPTION
**Pull request by fyrchik**
Implement basis for the oracle contract along with some tests.

### CODE REVIEW

**Review by roman-khimov — CHANGES_REQUESTED** [review]
Transaction verification part is still missing.

File: pkg/core/native/oracle.go [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+	oracleName       = "Oracle"
+)
+
+const (
+	maxURLLength      = 256
+	maxFilterLength   = 128
+	maxCallbackLength = 32
+	maxUserDataLength = 512
+
+	oracleRequestPrice = 5000_0000
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/native/oracle.go:46]: I think, that GAS prices are more readable this way, as it has 8 decimals.

File: pkg/core/native/oracle.go [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+	o.AddMethod(md, desc, true)
+
+	desc = newDescriptor("setOracleNodes", smartcontract.VoidType)
+	md = newMethodAndPrice(o.setOracleNodes, 0, smartcontract.AllowModifyStates)
+	o.AddMethod(md, desc, false)
+
+	desc = newDescriptor("verify", smartcontract.BoolType)
+	md = newMethodAndPrice(o.verify, 100_0000, smartcontract.NoneFlag)
+	o.AddMethod(md, desc, false)
+
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/native/oracle.go:118]: Fixed.

File: pkg/core/native/native_neo.go [hunk 3]
```diff
@@ -546,6 +547,18 @@ func (n *NEO) getNextBlockValidatorsInternal(bc blockchainer.Blockchainer, d dao
 	return pubs, nil
 }
 
+func (n *NEO) getCommitteeScriptHash(bc blockchainer.Blockchainer, d dao.DAO) (util.Uint160, error) {
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/native/native_neo.go]: Fixed.

File: pkg/core/native/native_neo.go [hunk 4]
```diff
@@ -546,6 +547,18 @@ func (n *NEO) getNextBlockValidatorsInternal(bc blockchainer.Blockchainer, d dao
...
+	pubs, err := n.GetCommitteeMembers(bc, d)
+	if err != nil {
+		return util.Uint160{}, err
+	}
+	script, err := smartcontract.CreateMajorityMultiSigRedeemScript(pubs)
+	if err != nil {
+		return util.Uint160{}, err
+	}
+	return hash.Hash160(script), nil
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/native/native_neo.go:699]: Fixed.

File: pkg/core/native/oracle.go [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+	emit.Syscall(w.BinWriter, interopnames.NeoNativeCall)
+	h := hash.Hash160(w.Bytes())
+
+	w.Reset()
+	emit.Int(w.BinWriter, 0)
+	emit.Opcode(w.BinWriter, opcode.NEWARRAY)
+	emit.String(w.BinWriter, "finish")
+	emit.Bytes(w.BinWriter, h.BytesBE())
+	emit.Syscall(w.BinWriter, interopnames.SystemContractCall)
+	oracleScript = w.Bytes()
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/native/oracle.go]: Fixed.

File: pkg/core/native/native_neo.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/core/native/native_neo.go]: I don't this being changed, `HighPriority` attribute also needs the committee script hash.

File: pkg/core/native_oracle_test.go [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+	}
+}
+
+func TestOracle_Request(t *testing.T) {
+	bc := newTestChain(t)
+	defer bc.Close()
+
+	orc := bc.contracts.Oracle
+	cs := getOracleContractState(orc.Hash)
+	ic := bc.newInteropContext(trigger.System, bc.dao, nil, nil)
```
Comment (roman-khimov) [inline pkg/core/native_oracle_test.go:91]: `trigger.Application`? You can put contract's state directly into the `bc.dao`, but regular txes are not run with `trigger.System`.

File: pkg/core/native/oracle.go [hunk 7]
```diff
@@ -357,6 +355,7 @@ func (o *Oracle) SetOracleNodes(ic *interop.Context, pubs keys.PublicKeys) error
 	}
 
 	sort.Sort(pubs)
+	o.nodes.Store(pubs)
```
Comment (roman-khimov) [inline pkg/core/native/oracle.go:374]: Transaction can fail.

File: pkg/core/native/oracle.go [hunk 8]
```diff
@@ -356,6 +367,11 @@ func (o *Oracle) SetOracleNodes(ic *interop.Context, pubs keys.PublicKeys) error
 
 	sort.Sort(pubs)
 	o.nodes.Store(pubs)
+	script, err := smartcontract.CreateMajorityMultiSigRedeemScript(pubs.Copy())
+	if err != nil {
+		return err
+	}
+	o.oracleHash.Store(hash.Hash160(script))
```
Comment (roman-khimov) [inline pkg/core/native/oracle.go:379]: And here too.

File: pkg/core/native/oracle.go [hunk 9]
```diff
@@ -0,0 +1,432 @@
...
+			return tx.Attributes[i].Value.(*transaction.OracleResponse)
+		}
+	}
+	return nil
+}
+
+func (o *Oracle) finish(ic *interop.Context, _ []stackitem.Item) stackitem.Item {
+	err := o.FinishInternal(ic)
+	if err != nil {
+		panic(err)
```
Comment (roman-khimov) [inline pkg/core/native/oracle.go:217]: This is worth adding a test for, callback errors should be forwarded into transaction errors.

File: pkg/core/native/oracle.go [hunk 10]
```diff
@@ -0,0 +1,432 @@
...
+func (o *Oracle) verify(ic *interop.Context, _ []stackitem.Item) stackitem.Item {
+	return stackitem.NewBool(ic.Tx.HasAttribute(transaction.OracleResponseT))
+}
+
+func (o *Oracle) getOriginalTxID(d dao.DAO, tx *transaction.Transaction) util.Uint256 {
+	for i := range tx.Attributes {
+		if tx.Attributes[i].Type == transaction.OracleResponseT {
+			id := tx.Attributes[i].Value.(*transaction.OracleResponse).ID
+			req, _ := o.GetRequestInternal(d, id)
+			return req.OriginalTxID
```
Comment (roman-khimov) [inline pkg/core/native/oracle.go:407]: I'm not sure about this btw, it's not covered at the moment and it's a little strange to me that is only goes one level deep.

**Review by roman-khimov — CHANGES_REQUESTED** [review]
Some simple tests for method parameter checks would also be nice to have.

File: pkg/core/native_oracle_test.go (hunk 6 above)
Comment (fyrchik) [reply pkg/core/native_oracle_test.go:91]: Fixed.

File: pkg/core/native/oracle.go (hunk 7 above)
Comment (fyrchik) [reply pkg/core/native/oracle.go:374]: Fixed.

File: pkg/core/native/oracle.go (hunk 8 above)
Comment (fyrchik) [reply pkg/core/native/oracle.go:379]: Fixed.

File: pkg/core/native/native_neo.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/native/native_neo.go]: Added commit.

File: pkg/core/native/oracle.go (hunk 10 above)
Comment (fyrchik) [reply pkg/core/native/oracle.go:407]: What do you mean by "one level deep"?

File: pkg/core/native/oracle.go (hunk 9 above)
Comment (fyrchik) [reply pkg/core/native/oracle.go:220]: Added a test for this case.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**